071-deprecate-bridge-openapi-2.md

```diff
@@ -0,0 +1,50 @@
+
+## Current situation
+
+Currently, the codebase provides two JSON files describing the HTTP endpoints exposed by the bridge.
+The [`openapiv2.json`](https://github.com/strimzi/strimzi-kafka-bridge/blob/main/src/main/resources/openapiv2.json) uses OpenAPI v2.
+The [`openapi.json`](https://github.com/strimzi/strimzi-kafka-bridge/blob/main/src/main/resources/openapi.json) uses OpenAPI v3.
+In reality, the `openapi.json` file is used internally to "load" the HTTP endpoints definition via the Vert.x Web OpenAPI component in order to build the web routes.
+The HTTP endpoints specification via OpenAPI is also used by the Vert.x Web OpenAPI component to validate parameters and body on the incoming HTTP requests.
+The `openapiv2.json` is just used to be returned as resource when an HTTP client issues a request on the `/openapi` HTTP endpoint.
+Exposing the OpenAPI specification via a dedicated HTTP endpoint is useful to external systems like API gateways or tools for API testing and for clients code auto-generation.
+
+## Motivation
+
+The bridge has been exposing the HTTP endpoints definition via the OpenAPI v2 specification for supporting external systems and tools still using Swagger.
+Internally, it has always been using the OpenAPI v3 to "load" the HTTP endpoints definition, build the corresponding web routes and validate the parameters and body on the incoming HTTP requests.
+The OpenAPI v2 specification can be considered obsolete with the latest [release](https://swagger.io/specification/v2/) happened 10 years ago.
+Most of the API gateways, clients and tools are now mostly supporting the OpenAPI v3 specification.
+Furthermore, every time there are changes in the HTTP endpoints definition, we need to keep the two JSON files in sync, because they are used both for different purposes as explained before.
+
+## Proposal
+
+The proposal is about deprecating the OpenAPI v2 specification support in the next Strimgi HTTP bridge 0.29.0 release and removing it in the future 0.30.0 release.
+To make the transition smoothly, the idea is to have two new HTTP endpoints:
+
+* `/openapi/v2`: still exposing the bridge HTTP endpoints definition with the OpenAPI v2 specification.
+* `/openapi/v3`: exposing the bridge HTTP endpoints definition with the OpenAPI v3 specification.
+
+During the deprecation period, with the 0.29.0 release, the HTTP endpoints definition will be available with both OpenAPI v2 and v3 specification on the two different above endpoints.
+Any HTTP request issued to the current `/openapi` endpoint will be forwarded to the `/openapi/v2` endpoint, still with the OpenAPI v2 specification.
+
+With the 0.30.0 release, the `/openapi/v2` will be handled to return the `404 Not Found` HTTP status code instead.
+Any HTTP request issued to the `/openapi` endpoint will be forwarded to the `/openapi/v3`, with the OpenAPI v3 specification.
```

Comment:
TBH, this seems like an overkill if you wanna do it just for a single release. I wonder if it would make sense to keep both for more releases - e.g. 4 (or perhaps gate it by time - e.g. remove v2 in first minor or major release in 2025). Or just warn that it will be replaced with v3 in 0.29 and replace it in 0.30.

But what you propose here seems like a lot of effort to pretend some compatibility.

Reply:
> Or just warn that it will be replaced with v3 in 0.29 and replace it in 0.30.

My first idea was exactly this one (I should have added to the rejected list) but I thought that it wasn't providing the time to the users to try querying the v3 spec. I was also not sure about "how to warn" that v2 was going away? Just doc? Or I should have looked for something specific at HTTP level?
So I came with the proposal of having a release with both endpoints to give time users to try the v3 and adapt and then removing v2 in the next one.

We could even stay for more releases but why do you think that with more releases it makes much sense? For the effort you see to do it?

Reply:
Well, one release could be easily only ~1-2 months. That is not much time to adapt. So to me, it does not really seem worth the effort. As I said, I would do it either for longer (which is clearly better, but I have no idea how much effort would that mean) or not at all.

Reply:
I changed on using a time-wise approach, so removing with the first minor or major release of 2025.